storage/google/cloud/storage/blob.py

```diff
@@ -404,14 +408,22 @@ def _get_download_url(self):
         :returns: The download URL for the current blob.
         """
         if self.media_link is None:
-            download_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            base_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            scheme, netloc, path, query, frag = urlsplit(base_url)
+            query = parse_qsl(query)
+            # Only add `generation' for synthesized URLs: 'mediaLink' from
+            # server presumably already encodes it.
             if self.generation is not None:
-                download_url += u'&generation={:d}'.format(self.generation)
-            if self.user_project is not None:
-                download_url += u'&userProject={}'.format(self.user_project)
-            return download_url
+                query.append(('generation', '{:d}'.format(self.generation)))
         else:
-            return self.media_link
+            scheme, netloc, path, query, frag = urlsplit(self.media_link)
+            query = parse_qsl(query)
+
+        if self.user_project is not None:
+            query.append(('userProject', self.user_project))
```

Comment:
Whether or not the server supplies us a `mediaLink` property, any given `Blob` instance (client-side) can have `user_project` set on it when instantiated:  if it does, then we have to pass the equivalent `userProject` query string parameter for all API requests (except in `create`, where it is invalid).

I'd rather use `parse_qsl' and `urlencode` to normalize the process of adding the needed query string parameters:  accreting them by hand is just asking for bugs (e.g., if / when the template changes).